File: app/src/main/java/org/bbqapp/android/view/BaseActivity.java
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 bbqapp
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package org.bbqapp.android.view;

import android.content.Context;
import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.view.View;
import android.view.ViewGroup;

import com.squareup.leakcanary.RefWatcher;

import org.bbqapp.android.Injector;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import javax.inject.Inject;

import butterknife.ButterKnife;
import dagger.ObjectGraph;

public class BaseActivity extends AppCompatActivity implements Injector {
    private ObjectGraph objectGraph;

    @Inject
    RefWatcher refWatcher;

    private void prepareAndInject(Context context) {
        Injector injector = (Injector) context;
        if (objectGraph == null && injector != null && injector.getObjectGraph() != null) {
            objectGraph = injector.getObjectGraph().plus(getModules().toArray());
            inject(this);
        }
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        // dagger
        prepareAndInject(getApplicationContext());
    }

    @Override
    public void setContentView(int layoutResID) {
        super.setContentView(layoutResID);

        // butterknife
        ButterKnife.bind(this);
    }

    @Override
    public void setContentView(View view) {
        super.setContentView(view);

        // butterknife
        ButterKnife.bind(this);
    }

    @Override
    public void setContentView(View view, ViewGroup.LayoutParams params) {
        super.setContentView(view, params);

        // butterknife
        ButterKnife.bind(this);
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();

        // dagger
        objectGraph = null;

        // butterknife
        ButterKnife.unbind(this);

        refWatcher.watch(this);
    }

    @Override
    public ObjectGraph getObjectGraph() {
        return objectGraph;
    }

    @Override
    public void inject(Object o) {
        objectGraph.inject(o);
    }

    protected List<Object> getModules() {
        return new ArrayList<>(Arrays.<Object>asList(new ActivityModule(this)));
    }
}

File: app/src/main/kotlin/org/bbqapp/android/view/BaseActivity.kt
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 bbqapp
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package org.bbqapp.android.view

import android.content.Context
import android.os.Bundle
import android.view.View
import android.view.ViewGroup
import butterknife.ButterKnife
import com.github.salomonbrys.kodein.KodeinInjector
import com.github.salomonbrys.kodein.android.appKodein
import com.squareup.leakcanary.RefWatcher
import com.trello.rxlifecycle.components.support.RxAppCompatActivity
import dagger.ObjectGraph
import org.bbqapp.android.Injector
import timber.log.Timber

open class BaseActivity : RxAppCompatActivity(), Injector {
    private val injector = KodeinInjector()
    private var objectGraph: ObjectGraph? = null

    private val refWatcher: RefWatcher by injector.instance()

    private fun prepareAndInject(context: Context) {
        val injector = context as? Injector
        if (objectGraph == null && injector != null && injector.objectGraph != null) {
            objectGraph = injector.objectGraph.plus(*Array(modules.size, { modules[it] }))
            inject(this)
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        // dagger
        prepareAndInject(applicationContext)

        // kodein
        injector.inject(appKodein())
    }

    override fun setContentView(layoutResID: Int) {
        super.setContentView(layoutResID)

        // butterknife
        ButterKnife.bind(this)
    }

    override fun setContentView(view: View) {
        super.setContentView(view)

        // butterknife
        ButterKnife.bind(this)
    }

    override fun setContentView(view: View, params: ViewGroup.LayoutParams) {
        super.setContentView(view, params)

        // butterknife
        ButterKnife.bind(this)
    }

    override fun onDestroy() {
        super.onDestroy()

        // dagger
        objectGraph = null

        // butterknife
        ButterKnife.unbind(this)

        refWatcher.watch(this)
    }

    @Deprecated(message = "Use Kodein instead")
    override fun getObjectGraph(): ObjectGraph? {
        return objectGraph
    }

    @Deprecated(message = "Use Kodein instead")
    override fun inject(o: Any) {
        try {
            objectGraph?.inject(o)
        } catch (e: IllegalArgumentException) {
            Timber.e(e, "Could not inject $localClassName.");
        }
    }

    @Deprecated(message = "Use Kodein instead")
    protected val modules: List<Any>
        get() = listOf(ActivityModule(this))
}